Userland/Services/WebContent/InputSource.h

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include "AK/Debug.h"
+#include <AK/HashTable.h>
+#include <AK/NonnullOwnPtr.h>
+#include <AK/String.h>
+
+namespace Web::WebDriver {
+
+// InputState has lists of InputSources, but the pointer input source constructor
+// depends on InputState, so the definitions are circular
+// forward declare InputState here
+class InputState;
+
+// https : // w3c.github.io/webdriver/#input-sources
+enum class InputSourceType { none,
+    key,
+    pointer,
+    wheel };
+// this enum is for the types of devices, which make up the subtype field of the PointerInputSource class
+enum class PointingDeviceSubtype { mouse,
+    pen,
+    touch };
+
+class InputSource {
+    // Each input source has an input id which is stored as a key in the input state map.
+    // input id is a UUID
+    AK::String input_id;
+
+    // create an input source given input state, type and optional subtype:
+    // FIXME: It probably makes more sense to designate this switch to factory
+    // that the Input State can call into
+    InputSource(AK::NonnullOwnPtr<InputState> /*input_state*/, InputSourceType /*type*/,
+        PointingDeviceSubtype /*subtype*/)
+    {
```

Comment:
This constructor is also implicitly private, meaning nobody can call it.

Looking at the algorithm, I think what we want here is:
- A private constructor, which does nothing.
- A public static factory method corresponding to ["create an input source"](https://w3c.github.io/webdriver/#dfn-create-an-input-source). Could be named `InputSource::create()` for short.
- Then, that runs the steps, and calls a method for "create a null input source" etc. Which could be a static `NullInputSource::create()` method, following that part of the spec, and the same for the other types.